### ISSUE TITLE
Metrics over any read: feature-space kinds and an explicit kind × domain rule (PR #20 Q5)

### ISSUE THREAD

**Issue by can-goodfire** [goodfire-ai/causalab#33]
## Context

PR #20's spec question Q5: metrics should not be forced to bind to an
`lm_head` read. A `top_k` over a featurizer's output — *which coordinates of
this learned subspace are largest?* — is a perfectly ordinary measurement, and
today it is inexpressible.

Reads already go through featurizers (corpus document `04_das_im.json` reads
`v_cf` through the `rot` subspace), so the value is there. The restriction is
purely in metric binding, and it is not even in the spec's checklist:

```python
# causalab/protocol/validate.py:218
if metric.kind != "kl" and of_site.component != "lm_head":
    # not in the §5 checklist: the token-space kinds name vocab
    # entries, which only lm_head produces (an interpretation this
    # loader commits to; surfaced in the PR notes)
```

That comment is right about *why* the check exists and honest about its
status: it is an interpretation the loader committed to, with no rule number
and no failing-document test. This issue turns it into an explicit
kind × domain rule, and adds the kinds the rule then makes room for.

## What exists today

| piece | file | state |
|---|---|---|
| kind vocabulary | `causalab/protocol/schema.py:150-159` `MetricKind` | 7 kinds, all token-space |
| value fields | `causalab/protocol/schema.py:164` `METRIC_FIELDS` | per-kind extra fields |
| token-string kinds | `causalab/protocol/schema.py:188` `TOKEN_COLUMN_METRIC_KINDS` | the 5 that resolve strings — **already a domain partition, unnamed as one** |
| the implicit check | `causalab/protocol/validate.py:218` | `component != "lm_head"` ⇒ refuse, no rule number |
| lowering | `causalab/neural/pytorch_hooks/metrics.py:202` `compute_metric` | `logits = _last_pos_logits(of_value).float()` — assumes vocab-shaped |
| one position only | `causalab/neural/pytorch_hooks/metrics.py:189` `_last_pos_logits` | refuses a multi-position read |
| capability derivation | `causalab/protocol/backend.py:43`, `:83` | `full_logits` from a saved full `lm_head` read **or** a `top_k`/`class_probs` metric — by kind, not by bound read |

Note the third row: `TOKEN_COLUMN_METRIC_KINDS` already partitions the kinds
by whether they resolve token strings. The domain split this issue needs is
almost the same partition — it just also has to place `kl` and `top_k`, which
are token-space but resolve no strings.

## Proposal

**Every metric kind declares a domain.**

| domain | kinds | binds to |
|---|---|---|
| `token` | `logit_diff`, `token_logit`, `cross_entropy`, `class_probs`, `match`, `kl`, `top_k` (vocab form) | a read whose site component is `lm_head` |
| `feature` | `norm`, `cosine`, `top_k` (feature form) | **any** read |

The domain table lives beside `METRIC_FIELDS` in `schema.py`; the binding
check moves out of `validate.py`'s ad-hoc branch into a numbered §5 rule with
its own failing document.

**The new kinds**, chosen to be the smallest set that makes feature-space
reads measurable without inventing an algebra:

| kind | fields | result per example |
|---|---|---|
| `norm` | `of` (+ optional `p`, default 2) | `‖f‖_p` |
| `cosine` | `of`, `target` (a read) | `cos(f, g)` |
| `top_k` (feature) | `of`, `k` | the k largest coordinates: indices + values |

`top_k` is deliberately **one kind serving both domains**, not two. Its
meaning is already "the k largest entries of the read's last axis"; over a
vocab read those entries decode to tokens, over a feature read they are
coordinate indices. The output shape differs (`tokens`/`probs` vs
`indices`/`values`) and the difference is decided by the bound read's
component — which the domain rule now makes explicit rather than implicit.

*Alternative considered and rejected:* a separate `feature_top_k` kind. It
duplicates the reduction to encode a fact the read already carries, and it
would leave two kinds to keep in step forever.

**`cosine` takes a second read**, exactly as `kl` does, and inherits `kl`'s
existing shape check: both reads must tap the same component
(`validate.py:236`), extended to "same component *and* same feature width" so
a cosine between a `k=8` subspace read and a full residual read refuses at
load instead of broadcasting at run time.

**Mean/variance reductions are not in this set.** A mean over examples is what
`select` already does (`docs/workflow_protocol.md` §2.3) and a mean over
*dims* is a featurizer's job. Adding them as metric kinds would put the same
reduction in three places.

**Outputs land in the same parquet path** as token metrics — no new writer,
no new save extension. `norm` and `cosine` are scalars per example; feature
`top_k` is a small struct per example, the shape `top_k` already produces.

**A capability trap, found while scoping this.** `full_logits` is derived
partly from the metric kind: `_FULL_VOCAB_METRICS = {"top_k", "class_probs"}`
(`causalab/protocol/backend.py:43`, used at `:83`). Because `top_k` is one kind
serving both domains, a *feature-space* `top_k` would inherit that requirement
and force a whole-vocab materialization for a measurement that never touches
the vocabulary — a document routed away from a vocab-parallel backend for no
reason. The derivation must become domain-aware: `top_k` requires
`full_logits` only when its bound read taps `lm_head`. This is the one place
the shared-kind decision costs something, and it is cheap to pay here.

## Implementation plan

1. **`causalab/protocol/schema.py`** — `norm` and `cosine` into `MetricKind`;
   `METRIC_FIELDS` entries; a `METRIC_DOMAINS` table mapping each kind to
   `"token"` | `"feature"` | `"either"` (`top_k`); `p` in
   `OPTIONAL_METRIC_FIELDS` with default `2`, materialized into the canonical
   form the way `match.mode` is.
2. **`causalab/protocol/validate.py`** — replace the `:218` branch with a
   numbered rule: a `token`-domain kind requires `of` to tap `lm_head`; a
   `feature`-domain kind accepts any component; `either` derives its form from
   the bound read. Extend the `kl` same-component check (`:236`) to cover
   `cosine`, and to compare resolved widths, not just component names. Add the
   rule to the module docstring's interpretations list.
3. **`causalab/neural/pytorch_hooks/metrics.py`** — split the entry point: the
   `_last_pos_logits(...).float()` normalization (`:211`) is token-space
   lowering, not universal. Feature kinds take the read value with its last
   axis intact. `norm` → `linalg.vector_norm`; `cosine` → normalized dot over
   the last axis; feature `top_k` → `topk` returning indices + values.
4. **`causalab/protocol/backend.py`** — `requires` consults the bound read's
   component for `top_k` instead of the kind alone (`:83`), so a feature
   `top_k` no longer demands `full_logits`.
5. **`docs/intervention_protocol.md`** — §2.10's table gains the two kinds and
   a **domain** column; the prose "metrics bind to lm_head reads" becomes the
   domain rule; §5's checklist gains the rule number; §8's capability table
   notes that feature metrics do not require `full_logits`.

## Tests

- `tests/protocol/test_validation_rules.py` — the one-failing-document test the
  current check never had: `logit_diff` over a `resid` read refuses with the
  new rule number; `norm` over a `resid` read loads; `cosine` between two reads
  of different widths refuses.
- `tests/neural/pytorch_hooks/test_metrics.py` — `norm`, `cosine` and feature
  `top_k` against hand-computed oracles on tiny-random; `top_k` over an
  `lm_head` read still returns decoded tokens (no regression in the shared
  kind).
- An end-to-end document computing `top_k` over the `rot` subspace read of the
  DAS corpus document: loads, validates, executes, matches the oracle.
- `tests/protocol/test_backend_routing.py` — a document whose only metric is a
  feature-space `top_k` does **not** require `full_logits`, while a vocab
  `top_k` still does. This is the regression the shared kind invites.
- `tests/protocol/test_canonical.py` — `norm`'s `p` default is materialized;
  **corpus digests unchanged** (new optional structure, no existing document
  touched).

## Non-goals

- **Per-position metric grids.** `compute_metric` still reduces exactly one
  position per example (`metrics.py:189`). An all-positions read plus a
  feature metric is still a saved tensor, not a table — a per-position metric
  needs a `position` column in the metric table and is separate work
  (`docs/test_migration.md`, "Per-position metric grids").
- **Cross-read arithmetic.** Differences of saved metrics stay post-hoc
  (§2.10). `cosine` takes two reads because a pairwise geometry has no other
  spelling — it is not a precedent for general binary metrics.
- **Mean/variance kinds** — see above.
- No change to the featurizer vocabulary, and no new save extension.

## Acceptance

- A document computing `top_k` over a subspace-featurizer read loads,
  validates, executes on tiny-random, and matches a hand-computed oracle.
- The kind × domain rule has a number in `docs/intervention_protocol.md` §5
  and a failing-document test in `tests/protocol/test_validation_rules.py` —
  replacing the unnumbered `validate.py:218` interpretation.
- `norm` and `cosine` are lowered and tested; `cosine` refuses mismatched
  widths at load.
- Feature-space metric output lands in the ordinary metric parquet table.
- `tests/protocol/corpus_digests.json` is byte-identical.

Depends on #20 (the protocol refactor).
